### PR TITLE
[DLS] get_access_control and access_control_query for SPO

### PR DESCRIPTION
**Pull request by timgrein**
## Closes https://github.com/elastic/enterprise-search-team/issues/4311

This PR adds the implementations for `get_access_control` and `access_control_query` for SPO. It also adjusts `site_groups` and `site_users` to use scrolling instead of fetching everything at once.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)

## Release Note

Sharepoint Online supports Document Level Security on Ingestion side.

### CODE REVIEW

File: connectors/sources/sharepoint_online.py [hunk 1]
```diff
@@ -983,6 +1005,43 @@ def _dls_enabled(self):
 
         return self._features.document_level_security_enabled()
 
+    def access_control_query(self, access_control):
+        return {"query": {"template": {"params": {"access_control": access_control}}}}
```
Comment (wangch079) [inline connectors/sources/sharepoint_online.py:1009]: Should this be a private method? e.g. `_access_control_query`, as this is called from a private method. Does it need to be present in the `BaseDataSource` class?

Comment (timgrein) [reply connectors/sources/sharepoint_online.py:1009]: No as we'll need that later on for the custom role provider to work. Every data source supporting DLS need to implement this.

File: connectors/sources/sharepoint_online.py [hunk 2]
```diff
@@ -983,6 +1005,43 @@ def _dls_enabled(self):
...
+
+                username = user.get("LoginName", None)
+                email = user.get("Email", None)
+                additional_fields = [
+                    value
+                    for value in [username, email]
+                    if value is not None and len(value) > 0
+                ]
+
+                access_control = list(set(groups + additional_fields + DEFAULT_GROUPS))
```
[the remark this answers is not in the document]
Comment (timgrein) [reply connectors/sources/sharepoint_online.py:1093]: That makes total sense, adjusted the implementation to prefix emails, usernames and groups.

File: connectors/sources/sharepoint_online.py [hunk 3]
```diff
@@ -619,6 +623,16 @@ async def users_and_groups_for_role_assignment(self, site_web_url, role_assignme
             # This can also mean "not found" so handling it explicitly
             return []
 
+    async def groups_for_user(self, site_web_url, user_id):
+        self._validate_sharepoint_rest_url(site_web_url)
+
+        url = f"{site_web_url}/_api/web/GetUserById('{user_id}')/groups"
+
+        try:
+            return await self._rest_api_client.fetch(url)
```
Comment (wangch079) [inline connectors/sources/sharepoint_online.py:632]: This one doesn't need to be `scroll`ed?

File: connectors/sources/sharepoint_online.py [hunk 4]
```diff
@@ -983,6 +1054,46 @@ def _dls_enabled(self):
...
+                groups = await self.client.groups_for_user(site_web_url, user.get("Id"))
+                groups = list(
+                    map(
+                        lambda group: _prefix_group(group.get("LoginName")),
+                        groups.get("value", []),
+                    )
+                )
+
+                username = user.get("LoginName", None)
+                email = user.get("Email", None)
```
Comment (wangch079) [inline connectors/sources/sharepoint_online.py:1086]:
```suggestion
                username = _prefix_user(user.get("LoginName", None))
                email = _prefix_email(user.get("Email", None))
```

File: connectors/sources/sharepoint_online.py (hunk 3 above)
Comment (timgrein) [reply connectors/sources/sharepoint_online.py:632]: AFAIU not necessarily as this is specific to the user and not the pages, but I may misunderstood the topology of SPO.

**Review by wangch079 — APPROVED** [review]

**Pull request merged**